src/node/internal/internal_dns_client.ts

```diff
@@ -0,0 +1,243 @@
+  Authority: Answer[];
+  Additional?: Answer;
+}
+
+export async function sendDnsRequest(
+  name: string,
+  type: string
+): Promise<SuccessResponse> {
+  // We are using cloudflare-dns.com and not 1.1.1.1 because of certificate issues.
+  // TODO(soon): Replace this when KJ certificate issues are resolved.
+  const server = new URL('https://cloudflare-dns.com/dns-query');
+  server.searchParams.set('name', name);
+  server.searchParams.set('type', type);
+
+  // syscall needs to be in format of `queryTxt`
+  // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+  const syscall = `query${type.at(0)?.toUpperCase()}${type.slice(1)}`;
+
+  let json: SuccessResponse | FailedResponse;
+  try {
+    const response = await fetch(server, {
+      headers: {
+        Accept: 'application/dns-json',
+      },
+      method: 'GET',
+    });
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    json = await response.json();
+  } catch {
+    throw new DnsError(name, errorCodes.BADQUERY, syscall);
+  }
+
+  if ('error' in json) {
+    throw new DnsError(name, errorCodes.BADRESP, syscall);
+  }
+
+  if (!json.Question.at(0)) {
+    // Some APIs depend on Question being existent.
+    throw new DnsError(name, errorCodes.BADRESP, syscall);
+  }
+
+  if (json.Answer?.at(0)?.name === '') {
+    throw new DnsError(name, errorCodes.NOTFOUND, syscall);
+  }
+
+  return json;
+}
+
+export function validateAnswer(
+  answer: unknown,
+  name: string,
+  query: string
+): asserts answer is Answer[] {
+  if (answer == null) {
+    throw new DnsError(name, errorCodes.NOTFOUND, query);
+  }
+}
+
+export type MX = {
+  exchange: string;
+  priority: number;
+};
+export function normalizeMx(name: string, answer: Answer): MX {
+  // eslint-disable-next-line prefer-const
+  let [priority, exchange]: string[] = answer.data.split(' ');
+  if (priority == null || exchange == null) {
+    throw new DnsError(name, errorCodes.BADRESP, 'queryMx');
+  }
+
+  // Cloudflare API returns "data": "10 smtp.google.com." hence
+  // we need to parse it. Let's play it safe.
+  if (exchange.endsWith('.')) {
+    exchange = exchange.slice(0, -1);
+  }
+
+  return {
+    exchange,
+    priority: parseInt(priority, 10),
+  };
+}
+
+export function normalizeCname({ data }: Answer): string {
+  // Cloudflare DNS returns "nodejs.org." whereas
+  // Node.js returns "nodejs.org" as a CNAME data.
+  if (data.endsWith('.')) {
+    return data.slice(0, -1);
+  }
+  return data;
+}
+
+export type CAA = {
+  critical: number;
+  issue?: string;
+  iodef?: string;
+  issuewild?: string;
+};
+export function normalizeCaa({ data }: Answer): CAA {
+  // CAA API returns "hex", so we need to convert it to UTF-8
+  const obj = { critical: 0 };
+  const record = dnsUtil.parseCaaRecord(data);
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error
+  obj[record.field] = record.value;
+  obj.critical = record.critical;
+  return obj;
```

Comment:
This would be better:

```
const record = dnsUtil.parseCaaRecord(data);
  const obj: CAA = { critical: record.critical };
  obj[record.field] = record.value;
  return obj;
  ```